pkgs/finalfusion-utils/default.nix

```diff
@@ -0,0 +1,90 @@
+{ stdenv
+, lib
+, callPackage
+, defaultCrateOverrides
+, fetchFromGitHub
+
+# Native build inputs
+, installShellFiles ? null # Available in 19.09 and later.
+, perl
+, pkgconfig
+
+# Build inputs
+, darwin
+, mkl
+, openblasCompat
+, gfortran
+
+# Build finalfusion-utils with MKL or OpenBLAS support. MKL support
+# requires an MKL derivation that exposes pkg-config files.
+, withMkl ? false
+, withOpenblas ? false
+}:
+
+assert !(withMkl && withOpenblas);
+
+let
+  src = fetchFromGitHub {
+    owner = "finalfusion";
+    repo = "finalfusion-utils";
+    rev = "0.9.0";
+    sha256 = "0ffwh3rbjn9cl8fdg9jaq035lkphbrhpxjm2dqvx46kzjn9550qm";
+  };
+  cargo_nix = callPackage ./finalfusion-utils.nix {
+    rootFeatures = [ "default" ]
+                   ++ lib.optional withOpenblas "openblas"
+                   ++ lib.optional withMkl "intel-mkl" ;
+  };
+in cargo_nix.workspaceMembers.finalfusion-utils.build.override {
+  crateOverrides = defaultCrateOverrides // {
+    finalfusion-utils = attr: rec {
+      inherit src;
+
+      pname = "finalfusion-utils";
+      name = "${pname}-${attr.version}";
+
```

Comment:
These lines seem a bit odd, `pname` is only used in `name`?

Reply:
Yes. Package normally have three attributes related to the name/version: `pname` (package name), `version`, and `name` (package name + version).

Normally you only define `pname` and `version` and `name` is generated automatically. `stdenv` has the following code for that:

```
name = "${attrs.pname}-${attrs.version}";
```

However, the Rust crates that were generated from `Cargo.lock` have a `pname` like `rust_finalfusion-utils`. I find this a bit ugly for something that is a standalone program, so I want to set the `pname` to `finalfusion-utils`.

Nix has a mechanism for overriding derivations. So, you can take any derivation (from e.g. `nixpkgs`) and make your own version by overriding the arguments (`override`) or attributes (`overrideAttr`). E.g. nixpkgs does not have Boost 1.58.0, but you could just produce it from the derivation for Boost 1.59.0 that is in nixpkgs by overriding the version and source:

```nix
boost159.overrideAttrs (oldAttrs: rec {
  name = "${oldAttrs.pname}-${version}";
  version = "1.58.0";

  src = fetchurl {
    url = "mirror://sourceforge/boost/boost_1_58_0.tar.bz2";
    sha256 = "1rfkqxns60171q62cppiyzj8pmsbwp1l8jd7p6crriryqd7j1z7x";
  };
});
```

One small downside of overriding is that the earlier definition of `name` is not applied automatically to overrides. So, for that reason I need to set `name` explicitly. Even though `pname` is not used elsewhere, the attribute is available (and used by some Nix tools):

```
% nix repl default.nix
nix-repl> finalfusion-utils.pname
"finalfusion-utils"
nix-repl> finalfusion-utils.name 
"finalfusion-utils-0.9.0"
```

Since this reply is turning into a monologue anyway ;), let me also show an example of argument overriding. `finalfusion-utils` defaults to OpenBLAS in `default.nix`. So, if I wanted to use MKL instead, one could just override the relevant arguments:

```
% nix-build -E 'with (import ./default.nix {}); finalfusion-utils.override { withMkl = true; withOpenblas = false; }'
/nix/store/7bm4s9m6h805722k28gcg921xab2mdjb-finalfusion-utils-0.9.0
```

Reply:
> Nix has a mechanism for overriding derivations. So, you can take any derivation (from e.g. nixpkgs) and make your own version by overriding the arguments (override) or attributes (overrideAttr). E.g. nixpkgs does not have Boost 1.58.0, but you could just produce it from the derivation for Boost 1.59.0 that is in nixpkgs by overriding the version and source:

This looks somewhat...hacky? Disguising the boost 1.58.0 package as 1.59.0..

> One small downside of overriding is that the earlier definition of name is not applied automatically to overrides. So, for that reason I need to set name explicitly. Even though pname is not used elsewhere, the attribute is available (and used by some Nix tools):

Thanks for explanations again!

Reply:
> This looks somewhat...hacky? Disguising the boost 1.58.0 package as 1.59.0..

Why? A lot of packages rarely change their build procedure between releases. In the nixpkgs repo the vast majority of package upgrades just bump the version and the hash. The benefit is that you don't have to maintain your own derivations and get all the improvements people make to the 1.59.0 derivation.

Reply:
(Alternatively, you could also use `fetchTarball` to get the nixpkgs revision that had 1.58.0 and use Boost from that.)

Reply:
No actual reason here, it just seemed odd to put `version = "1.58.0"` into something labelled as `1.59` ;)

Reply:
I have to admit I rarely use `overrideAttrs` (I do use `override` often). This particular example was from my Alpino derivation. Alpino is distributed as source + binaries, but you can't really compile it from source unless you have the proprietary SICStus Prolog interpreter. So, I just deliver the Boost version the Alpino binaries want ;).